Serializers for unmodifiable Java collections.

chermenin opened the pull request:
Added serializers for collections created via
Collections.unmodifiable...() and Collections.empty...() methods.

chermenin committed:
Serializers for unmodifiable and empty collections.

Added serializers for collections created via
Collections.unmodifiable...() and Collections.empty...() methods.

johnynek commented on chill-java/src/main/java/com/twitter/chill/java/UnmodifiableCollectionSerializer.java:
```diff
@@ -0,0 +1,81 @@
+ *
+ * @author <a href="mailto:alex@chermenin.ru">Alex Chermenin</a>
+ */
+public class UnmodifiableCollectionSerializer extends Serializer<Collection<?>> {
```
these are all basically the same. Can we share the code? Something like:

``` java
public abstract class UnmodifiableJavaCollectionSerializer<T, U> extends Serializer<T> {
    protected U getInner(T t);
    protected T build(U u);

    @Override
    public Collection<?> read(Kryo kryo, Input input, Class<T> type) {
        try {
            U c = (U) kryo.readClassAndObject(input);
            return build(c);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void write(Kryo kryo, Output output, T object) {
        try {
            U c = getInner(object);
            kryo.writeClassAndObject(output, c);
        } catch (RuntimeException e) {
            // Don't eat and wrap RuntimeExceptions because the ObjectBuffer.write...
            // handles SerializationException specifically (resizing the buffer)...
            throw e;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}
```

Then we can just implement the code to read the inner and wrap for each of them.

Also, could you possibly set your editor to 2 space indent, which is the style we use?

chermenin commented:
Great thought! Thanks for review.

chermenin committed:
Some refactoring.

Serializers of unmodifiable Java collections was refactored.
Deleted serializers for empty collections (Kryo has default serializators for it).
